Champions/Shaco/W.cs

```diff
@@ -0,0 +1,111 @@
+        }
+
+        public void OnStartCasting(IChampion owner, ISpell spell, IAttackableUnit target)
+        {
+        }
+
+        public void OnFinishCasting(IChampion owner, ISpell spell, IAttackableUnit target)
+        {
+            var fearduration = 0.5f + (0.25 * (spell.Level - 1));
+            var jackduration = 5.0f;
+            var apbonus = owner.Stats.AbilityPower.Total * 0.2f;
+            var damage = 35 + ((15 * (spell.Level - 1)) + apbonus);
+            var attspeed = 1 / 1.8f; // 1.8 attacks a second = .55... seconds per attack
+            var castrange = spell.SpellData.CastRange[0];
+            //TODO: Implement Fear buff
+            //var fearrange = 300;
+            var sightrange = 600;
+            var ownerPos = new Vector2(owner.X, owner.Y);
+            var spellPos = new Vector2(spell.X, spell.Y);
+
+            if (owner.WithinRange(ownerPos, spellPos, castrange))
+            {
+                IMinion m = AddMinion(owner, "ShacoBox", "ShacoBox", spell.X, spell.Y, sightrange);
+                AddParticle(owner, "JackintheboxPoof.troy", spell.X, spell.Y);
+
+                if (m.IsVisibleByTeam(owner.Team))
+                {
+                    try
+                    {
+                        if (!m.IsDead)
+                        {
+                            var units = GetUnitsInRange(m, sightrange, true);
+                            foreach (var value in units)
+                            {
+                                if (owner.Team != value.Team && value is IAttackableUnit && !(value is IBaseTurret) && !(value is IObjAnimatedBuilding))
+                                {
+                                    //TODO: Change TakeDamage to activate on Jack AutoAttackHit, not use CreateTimer, and make Pets use owner stats
+                                    m.SetTargetUnit(value);
+                                    m.AutoAttackTarget = value;
+                                    m.AutoAttackProjectileSpeed = 1450;
+                                    m.AutoAttackHit(value);
+                                    for (petTimeAlive = 0.0f; petTimeAlive < jackduration; petTimeAlive += attspeed)
+                                    {
+                                        CreateTimer(petTimeAlive, () => {
+                                            if (!value.IsDead && !m.IsDead)
+                                            {
+                                                value.TakeDamage(m, damage, DamageType.DAMAGE_TYPE_MAGICAL, DamageSource.DAMAGE_SOURCE_SPELL, false);
+                                            }
+                                        });
+                                    }
+                                }
+                            }
+                            CreateTimer(jackduration, () =>
+                            {
+                                if (!m.IsDead)
+                                {
+                                    m.Die(m);
```

Comment:
Doesn't m.Die should remove this minion? Or at least remove should call `Die` function?

Reply:
`Die` seems to also handle XP and some other things. The main problem is that it's not calling ot remove, and only `SetToRemove` which is not exactly remove the object (it sets some flag that I think not used well).